lib/ansible/galaxy/collection/gpg.py

```diff
@@ -0,0 +1,269 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2022, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""Signature verification helpers."""
+
+from ansible.errors import AnsibleError
+from ansible.module_utils.urls import open_url
+
+import os
+import subprocess
+import sys
+
+from dataclasses import dataclass, fields as dc_fields
+from functools import partial
+
+try:
+    # NOTE: It's in Python 3 stdlib and can be installed on Python 2
+    # NOTE: via `pip install typing`. Unnecessary in runtime.
+    # NOTE: `TYPE_CHECKING` is True during mypy-typecheck-time.
+    from typing import TYPE_CHECKING
+except ImportError:
+    TYPE_CHECKING = False
+
+if TYPE_CHECKING:
+    from ansible.utils.display import Display
+    from typing import Tuple, Iterator
+
+
+IS_PY310_PLUS = sys.version_info[:2] >= (3, 10)
+
+frozen_dataclass = partial(dataclass, frozen=True, **({'slots': True} if IS_PY310_PLUS else {}))
+
+
+def run_gpg_verify(
+    manifest_file,  # type: str
+    signature,  # type: str
+    keyring,  # type: str
+    display,  # type: Display
+):  # type: (...) -> Tuple[str, int]
+    status_fd_read, status_fd_write = os.pipe()
+
+    # running the gpg command will create the keyring if it does not exist
+    remove_keybox = not os.path.exists(keyring)
+
+    cmd = [
+        'gpg',
+        f'--status-fd={status_fd_write}',
+        '--verify',
+        '--batch',
+        '--no-tty',
+        '--no-default-keyring',
+        f'--keyring={keyring}',
+        '-',
+        manifest_file,
+    ]
+    cmd_str = ' '.join(cmd)
+    display.vvvv(f"Running command '{cmd}'")
+
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            pass_fds=(status_fd_write,),
+            encoding='utf8',
+        )
+    except (FileNotFoundError, subprocess.SubprocessError) as err:
+        raise AnsibleError(
+            f"Failed during GnuPG verification with command '{cmd_str}': {err}"
+        ) from err
+    else:
+        stdout, stderr = p.communicate(input=signature)
+    finally:
+        os.close(status_fd_write)
+
+    if remove_keybox:
+        try:
+            os.remove(keyring)
+        except OSError:
+            pass
```

Comment:
Consider using
```suggestion
        with contextlib.suppress(OSError):
            os.remove(keyring)
```
to better express the semantics.

Reply:
Nice, I haven't used `contextlib.suppress` before.